File: src/ccrouter/router.py
"""Decision pipeline: preset rules → message pick → heuristic → model."""

import time

from . import classifier, config, rules


class Decision:
    __slots__ = (
        "tier", "model", "rule", "score", "signals", "hops", "picked_text", "ms"
    )

    def __init__(self, tier, model, rule, score=0, signals=None, hops=0,
                 picked_text="", ms=0.0):
        self.tier = tier
        self.model = model
        self.rule = rule
        self.score = score
        self.signals = signals or []
        self.hops = hops
        self.picked_text = picked_text
        self.ms = ms

    def as_dict(self) -> dict:
        return {
            "tier": self.tier,
            "model": self.model,
            "rule": self.rule,
            "score": self.score,
            "signals": self.signals,
            "hops": self.hops,
        }


def decide(path: str, headers: dict, body: dict, raw_body_len: int,
           cfg: dict) -> "Decision | None":
    """Return a Decision for sentinel requests, or None for passthrough.

    Never raises: any internal error fails open to the MID tier (the
    sentinel is not a real model, so a routed request must get *some*
    rewrite).
    """
    started = time.time()
    if not rules.is_routable_path(path) or not isinstance(body, dict):
        return None
    if not rules.is_sentinel(body, cfg):
        return None
    try:
        decision = _decide_tier(headers, body, raw_body_len, cfg)
    except Exception as exc:  # fail open, always answer
        decision = Decision("mid", cfg["models"]["mid"], "fail_open",
                            signals=["error:%s" % type(exc).__name__])
    decision.ms = round((time.time() - started) * 1000, 2)
    return decision


def _decide_tier(headers: dict, body: dict, raw_body_len: int,
                 cfg: dict) -> Decision:
    models = cfg["models"]

    fixed = rules.subagent_fixed_tier(headers, cfg)
    if fixed:
        return Decision(fixed, models[fixed], "subagent_fixed")

    tier = rules.thinking_rule(body, cfg)
    if tier:
        return Decision(tier, models[tier], "thinking")

    text, hops, raw_last = classifier.pick_message(body.get("messages"), cfg)

    tier = rules.plan_mode_rule(raw_last, cfg)
    if tier:
        return Decision(tier, models[tier], "plan_mode", hops=hops,
                        picked_text=text)

    tier, score, signals = classifier.classify(text, cfg)
    rule = "heuristic" if hops == 0 else "continuation"
    if any(s.startswith("llm:") for s in signals):
        rule = "llm_tiebreak"

    # Very large contexts deserve at least the MID model.
    if rules.estimated_tokens(raw_body_len) > cfg["rules"]["long_context_tokens"]:
        floored = config.max_tier(tier, "mid")
        if floored != tier:
            signals.append("long_context_floor")
            tier = floored

    return Decision(tier, models[tier], rule, score=score, signals=signals,
                    hops=hops, picked_text=text)


def apply_param_fixups(body: dict, model: str, cfg: dict) -> list:
    """Adapt request params to the routed model's capabilities.

    Claude Code shapes params (effort level, thinking) for the model the
    *user* selected — behind the router that's the sentinel, so the session
    settings pass through unchecked. Fix what the target model would 400 on:
    unsupported effort levels are capped/stripped, thinking is stripped for
    models without it. Unknown models are left untouched.
    """
    fixups = []
    try:
        params = cfg.get("model_params", {}).get(model)
        if not params:
            return fixups

        output_config = body.get("output_config")
        if isinstance(output_config, dict) and "effort" in output_config:
            effort = output_config["effort"]
            allowed = params.get("allowed_efforts", [])
            if effort not in allowed:
                if allowed and params.get("effort_fallback"):
                    output_config["effort"] = params["effort_fallback"]
                    fixups.append("effort:%s->%s" % (effort, output_config["effort"]))
                else:
                    output_config.pop("effort")
                    fixups.append("effort:%s->dropped" % effort)
                if not output_config:
                    body.pop("output_config")

        if not params.get("allow_thinking", True) and "thinking" in body:
            body.pop("thinking")
            fixups.append("thinking:dropped")
            # Cascade: context-management strategies that require thinking
            # (clear_thinking_*) 400 once thinking is gone.
            context_management = body.get("context_management")
            if isinstance(context_management, dict) and isinstance(
                context_management.get("edits"), list
            ):
                kept = [
                    edit for edit in context_management["edits"]
                    if not (isinstance(edit, dict)
                            and str(edit.get("type", "")).startswith("clear_thinking"))
                ]
                if len(kept) != len(context_management["edits"]):
                    fixups.append("clear_thinking_edit:dropped")
                    if kept:
                        context_management["edits"] = kept
                    else:
                        body.pop("context_management")
    except Exception:
        return fixups
    return fixups

File: src/router.js
'use strict';

// Decision pipeline: preset rules -> message pick -> heuristic -> model.

const classifier = require('./classifier');
const config = require('./config');
const rules = require('./rules');

class Decision {
  constructor(tier, model, rule, opts = {}) {
    this.tier = tier;
    this.model = model;
    this.rule = rule;
    this.score = opts.score || 0;
    this.signals = opts.signals || [];
    this.hops = opts.hops || 0;
    this.pickedText = opts.pickedText || '';
    this.ms = opts.ms || 0;
  }

  asDict() {
    return {
      tier: this.tier,
      model: this.model,
      rule: this.rule,
      score: this.score,
      signals: this.signals,
      hops: this.hops,
    };
  }
}

// Return a Decision for sentinel requests, or null for passthrough. Never
// throws: any internal error fails open to the MID tier (the sentinel is not
// a real model, so a routed request must get *some* rewrite).
async function decide(path, headers, body, rawBodyLen, cfg) {
  const started = Date.now();
  if (!rules.isRoutablePath(path) || body === null || typeof body !== 'object' ||
      Array.isArray(body)) {
    return null;
  }
  if (!rules.isSentinel(body, cfg)) return null;

  let decision;
  try {
    decision = await decideTier(headers, body, rawBodyLen, cfg);
  } catch (err) {
    decision = new Decision('mid', cfg.models.mid, 'fail_open', {
      signals: [`error:${err && err.name ? err.name : 'Error'}`],
    });
  }
  decision.ms = Math.round((Date.now() - started) * 100) / 100;
  return decision;
}

async function decideTier(headers, body, rawBodyLen, cfg) {
  const models = cfg.models;

  const fixed = rules.subagentFixedTier(headers, cfg);
  if (fixed) return new Decision(fixed, models[fixed], 'subagent_fixed');

  let tier = rules.thinkingRule(body, cfg);
  if (tier) return new Decision(tier, models[tier], 'thinking');

  const [text, hops, rawLast] = classifier.pickMessage(body.messages, cfg);

  tier = rules.planModeRule(rawLast, cfg);
  if (tier) {
    return new Decision(tier, models[tier], 'plan_mode', { hops, pickedText: text });
  }

  let score, signals;
  [tier, score, signals] = await classifier.classifyAsync(text, cfg);
  let rule = hops === 0 ? 'heuristic' : 'continuation';
  if (signals.some((s) => s.startsWith('llm:'))) rule = 'llm_tiebreak';

  // Very large contexts deserve at least the MID model.
  if (rules.estimatedTokens(rawBodyLen) > cfg.rules.long_context_tokens) {
    const floored = config.maxTier(tier, 'mid');
    if (floored !== tier) {
      signals.push('long_context_floor');
      tier = floored;
    }
  }

  return new Decision(tier, models[tier], rule, {
    score, signals, hops, pickedText: text,
  });
}

// Adapt request params to the routed model's capabilities.
//
// Claude Code shapes params (effort level, thinking) for the model the *user*
// selected — behind the router that's the sentinel, so the session settings
// pass through unchecked. Fix what the target model would 400 on: unsupported
// effort levels are capped/stripped, thinking is stripped for models without
// it (plus dependent context-management edits). Unknown models are untouched.
function applyParamFixups(body, model, cfg) {
  const fixups = [];
  try {
    const params = (cfg.model_params || {})[model];
    if (!params) return fixups;

    const outputConfig = body.output_config;
    if (outputConfig && typeof outputConfig === 'object' && 'effort' in outputConfig) {
      const effort = outputConfig.effort;
      const allowed = params.allowed_efforts || [];
      if (!allowed.includes(effort)) {
        if (allowed.length && params.effort_fallback) {
          outputConfig.effort = params.effort_fallback;
          fixups.push(`effort:${effort}->${outputConfig.effort}`);
        } else {
          delete outputConfig.effort;
          fixups.push(`effort:${effort}->dropped`);
        }
        if (Object.keys(outputConfig).length === 0) delete body.output_config;
      }
    }

    if (params.allow_thinking === false && 'thinking' in body) {
      delete body.thinking;
      fixups.push('thinking:dropped');
      // Cascade: context-management strategies that require thinking
      // (clear_thinking_*) 400 once thinking is gone.
      const cm = body.context_management;
      if (cm && typeof cm === 'object' && Array.isArray(cm.edits)) {
        const kept = cm.edits.filter(
          (edit) => !(edit && typeof edit === 'object' &&
            String(edit.type || '').startsWith('clear_thinking')));
        if (kept.length !== cm.edits.length) {
          fixups.push('clear_thinking_edit:dropped');
          if (kept.length) cm.edits = kept;
          else delete body.context_management;
        }
      }
    }
  } catch {
    return fixups;
  }
  return fixups;
}

module.exports = { Decision, decide, applyParamFixups };
